src/miral/application_selector.cpp

```diff
@@ -21,243 +21,439 @@
+        } while (!can_select(*it));
+
+        on_raised(*start_it, *it, originally_selected);
+        return *it;
+    }
+
+    std::function<bool(T const&, T const&)> comparison_func;
+    std::function<bool(T const&)> can_select;
+    std::function<void(T const&, T const&, T const&)> on_raised;
+    std::function<void(T const&)> on_cancelled;
+    std::function<void(T&)> reset;
+
+    std::vector<T> focus_list;
+    /// The application that was selected when next was first called
+    T originally_selected;
+
+    /// The application that is currently selected.
+    T selected;
+
+    bool is_active = false;
+};
+
+}
+
+ApplicationSelector::ApplicationSelector(miral::WindowManagerTools const& _tools) :
+    tools{_tools},
+    window_cycler{std::make_unique<WindowCyclingBehavior<WeakApplication>>(
+        get_comparison_func(),
+        get_can_select_func(tools),
+        get_on_raised_func(tools),
+        get_on_cancelled_func(tools),
+        get_reset_func()
+    )}
+{
+}
 
 ApplicationSelector::~ApplicationSelector() = default;
 
 ApplicationSelector::ApplicationSelector(miral::ApplicationSelector const& other) :
-    tools{other.tools},
-    focus_list{other.focus_list},
-    originally_selected{other.originally_selected},
-    selected{other.selected},
-    active_window{other.active_window}
+    tools{other.tools}
 {
+    window_cycler = std::make_unique<WindowCyclingBehavior<WeakApplication>>(
+        get_comparison_func(),
+        get_can_select_func(tools),
+        get_on_raised_func(tools),
+        get_on_cancelled_func(tools),
+        get_reset_func(),
+        other.window_cycler->get_focus_list(),
+        other.window_cycler->get_originally_focused(),
+        other.window_cycler->get_focused(),
+        other.window_cycler->get_is_active()
+    );
 }
 
 auto ApplicationSelector::operator=(ApplicationSelector const& other) -> ApplicationSelector&
 {
     tools = other.tools;
-    focus_list = other.focus_list;
-    originally_selected = other.originally_selected;
-    selected = other.selected;
-    active_window = other.active_window;
+    window_cycler = std::make_unique<WindowCyclingBehavior<WeakApplication>>(
+        get_comparison_func(),
+        get_can_select_func(tools),
+        get_on_raised_func(tools),
+        get_on_cancelled_func(tools),
+        get_reset_func(),
+        other.window_cycler->get_focus_list(),
+        other.window_cycler->get_originally_focused(),
+        other.window_cycler->get_focused(),
+        other.window_cycler->get_is_active()
+    );
     return *this;
 }
 
 void ApplicationSelector::advise_new_app(Application const& application)
 {
-    focus_list.push_back(application);
+    window_cycler->add(application);
 }
 
 void ApplicationSelector::advise_focus_gained(WindowInfo const& window_info)
 {
     auto window = window_info.window();
     if (!window)
     {
+        mir::log_error("ApplicationSelector::advise_focus_gained: Could not find the window");
```

Comment:
Why is this an _error_? A WindowInfo with no window sounds like invariants have been broken. In which case, should we really be ignoring the problem and carrying on?

We ought to either:

1. Assume the invariant holds; or,
2. Report the program is in a bad state: `fatal_error()`